controllers/cloudstackmachine_controller_test.go

```diff
@@ -0,0 +1,104 @@
+/*
+Copyright 2022 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controllers_test
+
+import (
+	"github.com/golang/mock/gomock"
+	. "github.com/onsi/ginkgo/v2"
+	. "github.com/onsi/gomega"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	infrav1 "sigs.k8s.io/cluster-api-provider-cloudstack/api/v1beta2"
+	dummies "sigs.k8s.io/cluster-api-provider-cloudstack/test/dummies/v1beta2"
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1beta1"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+var _ = Describe("CloudStackMachineReconciler", func() {
+	Context("With machine controller running.", func() {
```

Comment:
is there some significance to these "context" strings that we should be aware of when developing/testing?

Reply:
[Organizing Specs With Container Nodes](https://onsi.github.io/ginkgo/#organizing-specs-with-container-nodes)
sort of explains it. Basically tests that share a certain setup will typically be within the same context.

Methods like `BeforeEach` will be shared among `Context` container nodes.

The nodes can be nested as well, so you could describe a situation kind of like:

``` Go
var _ = Describe("Person", func() {
	Context("On a nice sunny day", func() {
                Context("When wearing a sweater", func() {
                       It("Sweats a lot", func() {
                               ... Test code for a person wearing a sweater on a sunny day ...
                       })
                })
                Context("When wearing athletic shorts", func() {
                       It("Plays a game of soccer, then sweats a lot", func() {
                               ... Test code for a person wearing athletic shorts that plays soccer ...
                       })
                })
	})
```
	
So it's kind of a way to share test setup between individual tests.

In this case the machine controller is setup in the `BeforeEach` method.

``` Go
Ω(MachineReconciler.SetupWithManager(k8sManager)).Should(Succeed()) 
```